samza-api/src/main/java/org/apache/samza/system/IncomingMessageEnvelope.java

```diff
@@ -111,6 +111,13 @@ public SystemStreamPartition getSystemStreamPartition() {
     return systemStreamPartition;
   }
 
+  // used for elasticity to determine which elastic task should handle this envelope
+  public SystemStreamPartition getSystemStreamPartition(int elasticityFactor) {
+    Object envelopeKeyorOffset = key != null ? key : offset;
+    int keyBucket = Math.abs(envelopeKeyorOffset.hashCode() % elasticityFactor);
```

Comment:
10 % 3 is 1 
-10 % 3 is 2, 
so need to instead use Math.abs(envelopeKeyorOffset.hashCode()) % elasticityFactor ?